control-plane/grpc/src/tracing.rs

```diff
@@ -0,0 +1,199 @@
+            ])
+            .with_parent_context(context.clone())
+            .start(&tracer);
+
+        let context = context.with_span(span);
+        global::get_text_map_propagator(|propagator| {
+            propagator.inject_context(&context, &mut HeaderInjector(request.headers_mut()))
+        });
+        let service_clone = self.service.clone();
+        // we cannot simply use service_clone as self.service is ready but the service_clone may not
+        // be ready. Instead, swap them over.
+        let service = std::mem::replace(&mut self.service, service_clone);
+        trace_http_service_call(service, request, context)
+    }
+}
+
+/// Extract OpenTelemetry Spans from Http Headers
+#[derive(Clone, Default)]
+pub struct OpenTelServer {}
+impl OpenTelServer {
+    /// Return new `Self`
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+impl<S> tower::Layer<S> for OpenTelServer {
+    type Service = OpenTelServerService<S>;
+
+    fn layer(&self, service: S) -> Self::Service {
+        OpenTelServerService::new(service)
+    }
+}
+
+/// OpenTelemetry Service that extracts tracing spans from the Http Headers
+#[derive(Clone)]
+pub struct OpenTelServerService<S> {
+    service: S,
+}
+impl<S> OpenTelServerService<S> {
+    fn new(service: S) -> Self {
+        Self { service }
+    }
+}
+
+type TonicServerRequest = Request<tonic::transport::Body>;
+type TonicServerResponse = Response<tonic::body::BoxBody>;
+impl<S> tower::Service<TonicServerRequest> for OpenTelServerService<S>
+where
+    S: tower::Service<TonicServerRequest, Response = TonicServerResponse> + Send + Clone + 'static,
+    S::Future: Send,
+    S::Error: ToString,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = BoxedFuture<Self::Response, Self::Error>;
+
+    fn poll_ready(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        self.service.poll_ready(cx)
+    }
+
+    fn call(&mut self, request: TonicServerRequest) -> Self::Future {
+        let tracer = global::tracer_with_version("grpc-server", env!("CARGO_PKG_VERSION"));
+        let extractor = opentelemetry_http::HeaderExtractor(request.headers());
+
+        let parent_context =
+            global::get_text_map_propagator(|propagator| propagator.extract(&extractor));
+
+        let mut builder = tracer.span_builder(request.uri().to_string());
+        builder.parent_context = parent_context.clone();
+        builder.span_kind = Some(SpanKind::Server);
+
+        let span = tracer.build(builder);
+        let context = parent_context.with_span(span);
+
+        let service_clone = self.service.clone();
+        // we cannot simply use service_clone as self.service is ready but the service_clone may
+        // not be ready. Instead, swap them over.
+        let service = std::mem::replace(&mut self.service, service_clone);
```

Comment:
This pattern is used a couple of times. Would it be beneficial to encapsulate it in a function like "clone_from_service"? It may avoid having to remember why we have to do the replace.